Add read-only role page to documents #5783 (#6806)
opensearch-project/documentation-website: _security/access-control/users-roles.md

---

## Defining users

> You can define users using OpenSearch Dashboards, `internal_users.yml`, or the REST API. When creating a user, you can map users to roles using `internal_users.yml` or the REST API, but that feature is not currently available in OpenSearch Dashboards.

Reviewer: Is it true that you the mapping feature is not available in Dashboards? Because later we provide instructions on it.

> You can define users using OpenSearch Dashboards, `internal_users.yml`, or the REST API. When creating a user, you can map users to roles using `internal_users.yml` or the REST API. If you are using OpenSearch Dashboards to define users, mapping to roles is achived using **Mapped users** tab in `Security/Roles` section described further in [Map users to roles]({{site.url}}{{site.baseurl}}/security/access-control/users-roles/#map-users-to-roles).

Reviewer: ```suggestion
You can define users using OpenSearch Dashboards, `internal_users.yml`, or the REST API. When creating a user, you can map users to roles using `internal_users.yml` or the REST API. If you are using OpenSearch Dashboards to define users, mapping to roles is archived using **Mapped users** tab in `Security/Roles` section described further in [Map users to roles]({{site.url}}{{site.baseurl}}/security/access-control/users-roles/#map-users-to-roles).
```

> You can define users using OpenSearch Dashboards, `internal_users.yml`, or the REST API. When creating a user, you can map users to roles using `internal_users.yml` or the REST API. If you are using OpenSearch Dashboards to define users, mapping to roles is archived using **Mapped users** tab in `Security/Roles` section described further in [Map users to roles]({{site.url}}{{site.baseurl}}/security/access-control/users-roles/#map-users-to-roles).

Reviewer: ```suggestion
You can define users using OpenSearch Dashboards, `internal_users.yml`, or the REST API. When creating a user, you can map users to roles using `internal_users.yml` or the REST API. If you are using OpenSearch Dashboards to define users, mapping to roles is archived using **Mapped users** tab in `Security/Roles` section described further in [Map users to roles]({{site.url}}{{site.baseurl}}/security/access-control/users-roles/#mapping-users-to-roles).
```

See [Create user]({{site.url}}{{site.baseurl}}/security/access-control/api/#create-user).

## Defining roles

> Just like defining users, you can define roles using OpenSearch Dashboards, `roles.yml`, or the REST API. In addition, OpenSearch provides some predefined roles and a special read-only role, which you can read about later in this topic.

Reviewer 2: ```suggestion
Similar to defining users, you can define roles using OpenSearch Dashboards, `roles.yml`, or the REST API. In addition, OpenSearch provides some predefined roles and a special read-only role, which you can read about later in this topic.
```

 
- Using OpenSearch Dashboards
- Modifying the `yml` configuration files
- Using the Cluster Settings API 

The simplest way to get familiar with roles and role mappings is to use OpenSearch Dashboards. The interface simplifies creating roles and assigning those roles to users, with an easy-to-navigate workflow. 
{ .tip}

### Defining a basic read-only role

To create a basic read-only role that allows a user to access OpenSearch Dashboards, view existing dashboards and visualizations, and query different indexes, use the following permissions. These permissions give the user access to all tenants and indexes on the cluster.

#### Cluster permissions

For a user requiring read-only access to cluster-wide resources, such as visualizations or dashboards, add the `cluster_composite_ops_ro` permission to that user's role.

#### Index permissions

A user requiring access to view visualizations will also require access to the index used to create the visualization. To give a user read-only access to all indexes, specify all (`*`) under the **Index** dropdown and **Read** in **Index Permissions**.

#### Tenant permissions

> If you use tenants to split work between teams or projects, use the all (`*`) option followed by the **Read only** option, as demonstrated in the following image:

Reviewer 3: ```suggestion
If you use tenants to split work between teams or projects, use the all (`*`) option followed by the **Read only** option, as demonstrated in the following image.
```

![creating role]({{site.url}}{{site.baseurl}}/images/role_creation_read_only.png)

> After setting all permission types and defining the role, you can map the role directly to a user on the role's **Mapped users** tab. Select **Map users** and then choose a user to map to the role, as outlined in the following example:

Reviewer 3: ```suggestion
After setting all permission types and defining the role, you can map the role directly to a user on the role's **Mapped users** tab. Select **Map users** and then choose a user to map to the role, as shown in the following image.
```

See [Create role mapping]({{site.url}}{{site.baseurl}}/security/access-control/api/#create-role-mapping).

## Defining read-only roles

A read-only role grants users the ability to read data from the OpenSearch cluster, but not to modify or delete any data. The read-only role is useful for scenarios where you want to provide access to data for reporting, analysis, or visualization purposes, without allowing modifications to the data or the cluster itself. This helps maintain data integrity and prevents accidental or unauthorized changes.

As with any role in OpenSearch, read-only roles can be configured using the following three methods: 
- Using OpenSearch Dashboards.
- Modifying the `yml` configuration files.
- Using the Cluster Settings API. 

> The most user friendly approach for anyone getting familiar with roles and role mappings is to use OpenSearch Dashboards, as it is easy to navigate the creation of roles and assign those roles to users. 

Reviewer: ```suggestion
The most user-friendly approach for anyone getting familiar with roles and role mappings is to use OpenSearch Dashboards, where it's easy to navigate the creation of roles and assign those roles to users. 
```

{ .tip}

### Defining a basic read-only role

> To create a basic read-only role which allows access to OpenSearch Dashboards, view existing dashboards, visualizations and query different indexes, use the following permissions.

Reviewer: ```suggestion
To create a basic read-only role that allows users to access OpenSearch Dashboards, view existing dashboards and visualizations, and query different indexes, use the following permissions.
```

> These permissions will give the user access to all tenants and indexes on the cluster.

Reviewer: ```suggestion
The following permissions give the user access to all tenants and indexes on the cluster.
```

{: .note}

#### Cluster permission

> For the user needs read-only access to cluster-wide resources, such as visualization or dashboards, add the `cluster_composite_ops_ro` permission to that user's role.

Reviewer: ```suggestion
For the user that requires read-only access to cluster-wide resources, such as visualization or dashboards, add the `cluster_composite_ops_ro` permission to that user's role.
```

#### Index permission

> If the user needs access to view visualizations, they will also need access to the index used to create the visualization. To give the user read-only access to all indexes, specify all (`*`) under the **Index** drop-down, and **Read** in **Index Permissions**.

Reviewer: ```suggestion
If the user requires access to view visualizations, they will also require access to the index used to create the visualization. To give the user read-only access to all indexes, specify all (`*`) under the **Index** drop-down, and **Read** in **Index Permissions**.
```

#### Tenant permissions

If you use tenants to split work between different teams or projects, use the all (`*`) option followed by the **Read only** option.

![creating role]({{site.url}}{{site.baseurl}}/images/role_creation_read_only.png)

> After all permission types are set and the role is created, you can directly map the role to a user by going to **Mapped users** tab in roles. Select **Map users** and select the user to map to this role. 

Reviewer: ```suggestion
After all permission types are set and the role is defined, you can map the role directly to a user in the **Mapped users** tab in roles. Select **Map users** and select the user to map to this role. 
```

Reviewer: Is _roles_ a section in the UI? Then it should be capitalized. But is "in roles" really needed?

![mapping users]({{site.url}}{{site.baseurl}}/images/mapping-users.png)

### OpenSearch Dashboards `readonly_mode`

> OpenSearch Dashboards `readonly_mode` functionality is used to give the user access to only the `Dashboards` UI, removing all other elements of UI from the view.

Reviewer: ```suggestion
The OpenSearch Dashboards `readonly_mode` functionality is used to give the user access to the `Dashboards` UI only, removing all other UI elements from the view.
```

> To configure this, add the following line `opensearch_dashboards.yml` file:

Reviewer: ```suggestion
To configure this role, add the following line `opensearch_dashboards.yml` file:
```

```opensearch_security.readonly_mode.roles: [new_role]```

If the role mapped to the user has additional privileges or the user is mapped to other roles, giving them write access to indexes, this access will not be allowed using OpenSearch Dashboard. Direct data access to OpenSearch using curl or API is still allowed, as OpenSearch Dashboards is not involved in this communication. 

If the user is mapped to the `readonly_mode` role, all other elements of the UI will be removed, except for `Dashboards`. In the following comparison, the left view shows the screen from the perspective of a user mapped to a role in`readonly_mode`. On the right, the user is given a standard view.

![compare read only mode]({{site.url}}{{site.baseurl}}/images/compare_read_only_mode.png)

> Mapping the user to only `readonly_mode` role does not give permissions to view relevant indexes or allow the user to view the existing dashboards. Read access to indexes and dashboards require separate permissions.

Reviewer: ```suggestion
Mapping the user to only the `readonly_mode` role does not give them permissions to view relevant indexes or allow them to view the existing dashboards. Read access to indexes and dashboards require separate permissions.
```

{: .note }

If the user is also mapped to any roles listed under `plugins.security.restapi.roles_enabled` in `opensearch.yml`, for example `all_access` or `security_rest_api_access`, then `readonly_mode` is ignored, giving the user access to standard UI elements.

### Additional permissions

> If you need access to additional permissions while using the `read_only` role,  such as alerting and anomaly detection modules, check out the existing roles, such as `alerting_read_access` and `anomaly_read_access`.

Reviewer: ```suggestion
If the user requires access to additional permissions in addition to the `read_only` role,  such as alerting and anomaly detection tasks, you can use the existing predefined roles, such as `alerting_read_access` and `anomaly_read_access`.
```

See [Create role mapping]({{site.url}}{{site.baseurl}}/security/access-control/api/#create-role-mapping).

## Read-only roles

As with any role in OpenSearch, read-only roles can be configured using the following three methods: 
- Modifying the `yml` configuration files.
- Using the Cluster Settings API. 
- Using OpenSearch Dashboards. 

The most user friendly approach for anyone getting familiar with Roles and role mappings is to use OpenSearch Dashboards, as it is easy to navigate the creation of roles and assign those roles to users. The basic steps of creating roles, mapping, and users can be found in the [User and roles documentation](https://opensearch.org/docs/latest/security/access-control/users-roles/).

### Creating a basic read-only role

> To create a basic read-only role which allows access to OpenSearch Dashboards, view existing dashboards, visualizations and query different indexes, use one the following permissions. 

Reviewer 2: I'm assuming that in the **Create user** view in Dashboards that we would use all of these permissions for the `read_only` role. Is that correct?

```suggestion
To create a basic read-only role which allows access to OpenSearch Dashboards, view existing dashboards, visualizations and query different indexes, use the following permissions.
```

For more detailed summaries of the permissions for each role, reference their action groups against the descriptions in [Default action groups]({{site.url}}{{site.baseurl}}/security/access-control/default-action-groups/).

## Example 

> The following example demonstrate how you might set up a bulk access role in OpenSearch Dashboards.

Reviewer: ```suggestion
The following example demonstrates how you can set up a bulk access role in OpenSearch Dashboards.
```
